packages/runtime/src/ctx.ts

```diff
@@ -0,0 +1,247 @@
+import { AssetHolding } from "algosdk";
+
+import { Runtime } from ".";
+import { RUNTIME_ERRORS } from "./errors/errors-list";
+import { RuntimeError } from "./errors/runtime-errors";
+import { AccountAddress, AlgoTransferParam, AssetModFields, AssetTransferParam, SSCAttributesM, StoreAccountI } from "./types";
+
+export class Ctx {
+  private readonly runtime: Runtime;
+
+  constructor (runtime: Runtime) {
+    this.runtime = runtime;
+  }
+
+  // verify 'amt' microalgos can be withdrawn from account
+  assertMinBalance (amt: number, address: string): void {
+    const account = this.getAccount(address);
+    if ((account.amount - amt) < account.minBalance) {
+      throw new RuntimeError(RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE, {
+        amount: amt,
+        address: address
+      });
+    }
+  }
+
+  // verifies assetId is not frozen for an account
+  assertAssetNotFrozen (assetIndex: number, address: AccountAddress): void {
+    const assetHolding = this.getAssetHolding(assetIndex, address);
+    if (assetHolding["is-frozen"]) {
+      throw new RuntimeError(RUNTIME_ERRORS.TRANSACTION.ACCOUNT_ASSET_FROZEN, {
+        assetId: assetIndex,
+        address: address
+      });
+    }
+  }
+
+  /**
+   * Fetches account from `runtime.ctx`
+   * @param address account address
+   */
+  getAccount (address: string): StoreAccountI {
+    const account = this.runtime.ctx.state.accounts.get(address);
```

Comment:
Can you please explain why we're using `runtime.ctx` here? We are in `Ctx` class but we are operating on `runtime.ctx.state` here. Shouldn't this be `this.state` (which would be a deepCopy of `runtime.store`)?

Reply:
Because it is not required to make a deepcopy of a deepcopy.
think about the flow, first we make a deepcopy and send it to interpreter(interpreter works on runtime.ctx) if the execution fails at some point we don't update the runtime.store.

same is the case with processTransaction, we process transactions on deepcopy of store, if it fails we don;t update store with ctx, else we update store.

Reply:
let me know it that makes sense.

Reply:
No i understand about the deepCopy, i am talking about the intuitiveness here (using  `runtime.ctx.state`) inside a `Ctx` class.
When i was talking about using `Ctx` as a class, i was referring to `runtime.ctx` only (not `ctxOp`). Basically having `runtime.ctx = new Ctx(deepCopy(this.store), tx, gtxs, args..)`. Class `Ctx` would be of type `Context` in `/runtime/types` only. Interpreter would have used this class functions (instead of having local interpreter functions). This way the separation b/w `ctx & store` would be better also.

For here, maybe during initialization we can do `this.state = runtime.ctx.state` in the constructor? Otherwise it seems a little confusing to me. We can keep as it is if you think otherwise.

Reply:
ok, do you mean ``` export interface Context {
  state: State
  tx: Txn // current txn
  gtxs: Txn[] // all transactions
  args: Uint8Array[]
  transferAlgo()
  transferAsset()
}```

we will have these in types and we will have this class as Context. right ?
yes, this seems better.

ctx = deepCopy(this.store)
yes in constructor we can also new Ctx(ctx, tx, gtxs, args..)
use this.state = ctx
this.tx = tx
like this

Reply:
> we will have these in types and we will have this class as Context. right ?

Exactly. Instead of having `runtime.ctx` as an obj, we can convert that to a class (which we can define in a new file, as done here). So this:
```
this.ctx = {
  state: cloneDeep(this.store), // state is a deep copy of store
  tx: tx,
  gtxs: gtxs,
  args: []
};
 ```   
 would update to:
 ```
 this.ctx = new Ctx(cloneDeep(this.store), tx, gtxs, [])
 ```
 And all ctx related functions would be in `Ctx` class.
 
 If you like it then we can update. Your call. 
 Otherwise we can just update the constructor here like you said above.

Reply:
yes, i like it, updating to suggested one